src/ImmutableProperties.php

```diff
@@ -0,0 +1,120 @@
+<?php
+
+declare(strict_types=1);
+
+namespace Codelicia\Immutable;
+
+use ReflectionObject;
+use ReflectionProperty;
+use ReflectionType;
+use RuntimeException;
+use function array_key_exists;
+use function class_implements;
+use function get_class;
+use function gettype;
+use function in_array;
+use function interface_exists;
+use function is_object;
+use function spl_object_hash;
+use function sprintf;
+
+trait ImmutableProperties
+{
+    /** @return string[][]|ReflectionProperty[][] */
+    protected function &reflectionProperty(): array
+    {
+        static $reflectionProperty = [];
+
+        if (! array_key_exists(spl_object_hash($this), $reflectionProperty)) {
+            $reflectionProperty[spl_object_hash($this)] = [];
+        }
+
+        return $reflectionProperty[spl_object_hash($this)];
+    }
+
+    public function __construct()
+    {
+        $this->init();
+    }
+
+    public function init(): void
+    {
+        $reflection = new ReflectionObject($this);
+        $store = &$this->reflectionProperty();
+        $defaultProperties = $reflection->getDefaultProperties();
+
+        foreach ($reflection->getProperties(ReflectionProperty::IS_PUBLIC) as $reflectionProperty) {
```

Comment:
Why only `public` properties? This approach can work with any kind of property.

Reply:
It will work indeed! but I think it is conflicting with the visibility purpose. When I say something is private, I want to have more control over it, so I should be creating code that deals with that manually IMHO

Reply:
Although It makes sense for `protected` visibility IMO

Reply:
Nah! Changed my mind. I added a `visibilities()` methods to list the ones that shall be affected by the trait. So one can customize it if needed.